### PR TITLE
Deflate message support

### PR DESCRIPTION
**Pull request by ergurjeet**
Deflate message before publishing it to a queue and inflate it back before delivering it to the consumer.
It adds  `Content-Encoding: inflate` header to a message to identify the inflated message at the receivers end.

### CODE REVIEW

File: src/Queue.js [hunk 1]
```diff
@@ -59,7 +70,7 @@ export default class SqsQueue {
     };
 
     const finalMessage = {
-      MessageBody: JSON.stringify(message),
+      MessageBody: JSON.stringify(msgBody),
```
Comment (normrider2001) [inline src/Queue.js]: I think we should send raw text on a compressed message.

File: src/util.js [hunk 2]
```diff
@@ -28,6 +31,25 @@ export function safeEmit(q, eventName, arg) {
   }
 }
 
+
+export function compressMessage(message, compression) {
+  if (compression === true || compression.encoding === supportedCompression) {
+    return {
+      headers: { 'Content-Encoding': supportedCompression },
+      body: zlib.deflateSync(JSON.stringify(message)),
```
Comment (normrider2001) [inline src/util.js]: We should probably use the async call.

Also if we are sending text we might need to Base64 encode to send over SQS.  I am assuming what comes out of Zlib deflate is a byte array.

File: src/util.js [hunk 3]
```diff
@@ -45,16 +67,24 @@ export function messageHandlerFunc(context, sqsQueue, handler) {
...
+      }
     } catch (error) {
       logger.error('Failed to parse SQS Body as JSON', errorWrap(error));
       Object.defineProperty(error, ALREADY_LOGGED, { value: true, enumerable: false });
       safeEmit(sqsQueue.queueClient, 'error', callInfo);
       throw error;
     }
     try {
-      await handler(messageContext, parsedMessage, rest);
+      await handler(messageContext, decodedMessage || parsedMessage, decodedRest || rest);
```
Comment (normrider2001) [inline src/util.js]: parsedMessage and rest should be the only variable passed in here.  The || makes it confusing.

File: src/util.js [hunk 4]
```diff
@@ -45,16 +67,24 @@ export function messageHandlerFunc(context, sqsQueue, handler) {
     const errorWrap = context.service?.wrapError || context.gb?.wrapError || (e => e);
 
     let parsedMessage;
+    let decodedMessage;
+    let decodedRest;
     try {
+      const contentEncoding = rest.MessageAttributes?.['Content-Encoding']?.StringValue;
       parsedMessage = JSON.parse(Body);
+      if (contentEncoding === supportedCompression) {
```
Comment (normrider2001) [inline src/util.js:86]: If it's compressed { decompress, decode, parse  } , else { parse }

Only use parsedMessage

**Review by normrider2001 — CHANGES_REQUESTED** [review]
I don't think a compressed message should have JSON as the final message. It should just be text.

File: src/util.js (hunk 3 above)
Comment (ergurjeet) [reply src/util.js]: We need to retain both copies so we can publish it to the `*_unprocessable` queue if rejected by a service. However, we can definitely clean it up a bit

**Review by normrider2001 — APPROVED** [review]

**Pull request merged**